src/graphnet/data/datamodule.py

```diff
@@ -0,0 +1,429 @@
+            self._train_dataset, SQLiteDataset
+        ):
+            self._train_dataset._close_connection()
+
+        if hasattr(self, "_val_dataset") and isinstance(
+            self._val_dataset, SQLiteDataset
+        ):
+            self._val_dataset._close_connection()
+
+        if hasattr(self, "_test_dataset") and isinstance(
+            self._test_dataset, SQLiteDataset
+        ):
+            self._test_dataset._close_connection()
+
+        return
+
+    def _create_dataloader(
+        self, dataset: Union[Dataset, EnsembleDataset]
+    ) -> DataLoader:
+        """Create a DataLoader for the given dataset.
+
+        Args:
+            dataset (Union[Dataset, EnsembleDataset]): The dataset to create a DataLoader for.
+
+        Returns:
+            DataLoader: The DataLoader configured for the given dataset.
+        """
+        if dataset == self._train_dataset:
+            dataloader_args = self._train_dataloader_kwargs
+        elif dataset == self._val_dataset:
+            dataloader_args = self._validation_dataloader_kwargs
+        elif dataset == self._test_dataset:
+            dataloader_args = self._test_dataloader_kwargs
+        else:
+            raise ValueError(
+                "Unknown dataset encountered during dataloader creation."
+            )
+
+        if dataloader_args is None:
+            raise AttributeError("Dataloader arguments not provided.")
+
+        return DataLoader(dataset=dataset, **dataloader_args)
+
+    def _validate_dataset_class(self) -> None:
+        """Sanity checks on the dataset reference (self._dataset).
+
+        Checks whether the dataset is an instance of SQLiteDataset,
+        ParquetDataset, or Dataset. Raises a TypeError if an invalid dataset
+        type is detected, or if an EnsembleDataset is used.
+        """
+        print(self._dataset, "Dataset\n")
+        print(
+            f"Type of self._dataset before validation check: {type(self._dataset)}"
+        )
+        # if type(self._dataset) not in [SQLiteDataset, ParquetDataset, Dataset]:
+        #     raise TypeError(
+        #         "dataset_reference must be an instance of SQLiteDataset, ParquetDataset, or Dataset."
+        #     )
+        if isinstance(self._dataset, EnsembleDataset):
+            raise TypeError(
+                "EnsembleDataset is not allowed as dataset_reference."
+            )
+
+    def _validate_dataset_args(self) -> None:
+        """Sanity checks on the arguments for the dataset reference."""
+        if isinstance(self._dataset_args["path"], list):
+            if self._selection is not None:
+                try:
+                    # Check that the number of dataset paths is equal to the
+                    # number of selections given as arg.
+                    assert len(self._dataset_args["path"]) == len(
+                        self._selection
+                    )
+                except AssertionError:
+                    raise ValueError(
```

Comment:
You can shorten the `ValueError` lines by breaking it into multiple smaller strings. Like so:

```python
raise ValueError("The number of dataset paths" 
                 f" ({len(self._dataset_args['path'])})" 
                 " does not match the number of" 
                 f" selections ({len(self._selection)}).")
```